File: 11_动态规划/力扣加加/爬楼梯型/174. 地下城游戏.ts
// 编写一个函数来计算确保骑士能够拯救到公主所需的最低初始健康点数。
// 反向DP
// 状态定义：dp[i][j]表示从[i,j]到终点需要的最小血量，dp[0][0]就是最小初始血量
// 状态转移：1. 如果dungeon[i][j] == 0，那么，dp[i][j] = min(dp[i+1][j], dp[i][j+1])
//          2. 如果dungeon[i][j] < 0，那么，dp[i][j] = min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j]
//          3. 如果dungeon[i][j] > 0，那么，dp[i][j] = max(1, min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j])
// 所以，三种情况可以统一成一种dp[i][j] = max(1, min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j])
// 处理边界：dp[m-1][n-1] = max(1, 1-dungeon[m-1][n-1])，右边和下边，相临的元素只有一个，特殊处理一下。

function calculateMinimumHP(dungeon: number[][]): number {
  const m = dungeon.length
  const n = dungeon[0].length
  const dp = Array.from<number, number[]>({ length: m }, () => Array(n).fill(Infinity))

  dp[m - 1][n - 1] = Math.max(1, 1 - dungeon[m - 1][n - 1])
  for (let i = m - 2; i >= 0; i--) {
    dp[i][n - 1] = Math.max(1, dp[i + 1][n - 1] - dungeon[i][n - 1])
  }
  for (let j = n - 2; j >= 0; j--) {
    dp[m - 1][j] = Math.max(1, dp[m - 1][j + 1] - dungeon[m - 1][j])
  }

  for (let i = m - 2; i >= 0; i--) {
    for (let j = n - 2; j >= 0; j--) {
      dp[i][j] = Math.max(1, Math.min(dp[i + 1][j], dp[i][j + 1]) - dungeon[i][j])
    }
  }

  // console.table(dp)

  return dp[0][0]
}

console.log(
  calculateMinimumHP([
    [-2, -3, 3],
    [-5, -10, 1],
    [10, 30, -5],
  ])
)

export {}

File: 11_动态规划/dp分类/路径dp/174. 地下城游戏.py
# 计算确保骑士能够拯救到公主所需的最低初始健康点数
from functools import lru_cache
from typing import List


DIR2 = ((0, 1), (1, 0))


class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        @lru_cache(None)
        def dfs(row: int, col: int) -> int:
            if (row, col) == (ROW - 1, COL - 1):
                return max(1, 1 - dungeon[row][col])

            nextMin = int(1e20)
            for dr, dc in DIR2:
                nr, nc = row + dr, col + dc
                if 0 <= nr < ROW and 0 <= nc < COL:
                    nextMin = min(nextMin, dfs(nr, nc))
            return max(1, nextMin - dungeon[row][col])

        ROW, COL = len(dungeon), len(dungeon[0])
        res = dfs(0, 0)
        dfs.cache_clear()
        return res

